File: array/18.py
from typing import List

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums) < 4: return []
        nums.sort()
        x = 0
        ret = []
        while x <= len(nums) - 4:
            while x > 0 and x <= len(nums) - 4 and nums[x] == nums[x - 1]:
                x += 1
                continue
            y = x + 1
            while y <= len(nums) - 3:
                while y > x + 1 and y <= len(nums) - 3 and nums[y] == nums[y - 1]:
                    y += 1
                low = y + 1
                high = len(nums) - 1
                while low < high:
                    sum_ = nums[x] + nums[y] + nums[low] + nums[high]
                    if sum_ > target:
                        high -= 1
                        while low < high and nums[high] == nums[high + 1]:
                            high -= 1
                    elif sum_ < target:
                        low += 1
                        while low < high and nums[low] == nums[low - 1]:
                            low += 1
                    else:
                        ret.append([nums[x], nums[y], nums[low], nums[high]])
                        high -= 1
                        low += 1
                        while low < high and nums[high] == nums[high + 1]:
                            high -= 1
                        while low < high and nums[low] == nums[low - 1]:
                            low += 1
                    
                y += 1
            x += 1
        return ret


def main():
    nums = [1,-2,-5,-4,-3,3,3,5]
    target = -11
    print(Solution().fourSum(nums, target))

if __name__ == "__main__":
    main()

File: leetcode-cpp/18.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        if (nums.size() < 4) return vector<vector<int>>();
        vector<vector<int>> ret;
        sort(nums.begin(), nums.end());
        for (int base_one = 0; base_one < nums.size()-3; base_one++){
            if(base_one > 0 && nums[base_one] == nums[base_one-1])
                continue;
            for (int base_two = base_one + 1; base_two < nums.size()-2; base_two++){
                if (base_two > base_one + 1 && nums[base_two] == nums[base_two-1])
                    continue;
                int min_sum = nums[base_one] + nums[base_two] + nums[base_two+1] + nums[base_two+2];

                // 这一部分的计算非常节约时间
                if (target < min_sum)
                    continue;
                int max_sum = nums[base_one] + nums[base_two] + nums[nums.size()-1] + nums[nums.size()-2];
                if (target > max_sum)
                    continue;


                int i = base_two + 1, j = nums.size() - 1, temp;
                while(i < j) {
                    temp = nums[base_one] + nums[base_two] + nums[i] + nums[j];
                    bool l = false, r = false;
                    if (temp == target){
                        ret.push_back(vector<int>({nums[base_one], nums[base_two], nums[i], nums[j]}));
                        l = true;
                        r = true;
                    }else if(temp < target){
                        l = true;
                    }else{
                        r = true;
                    }
                    if (l) {
                        i++;
                        while(i < j && nums[i] == nums[i-1]) i++;
                    }
                    if (r){
                        j--;
                        while(i < j && nums[j] == nums[j+1]) j--;
                    }
                }
            }
        }
        return ret;
    }
};

int main(){
    Solution s;
    vector<int> nums({1,-2,-5,-4,-3,3,3,5});
    int target = -11;
    vector<vector<int>> ret = s.fourSum(nums, target);
    for (int i = 0; i < ret.size(); i++){
        for (int j = 0; j < ret[i].size(); j++){
            cout << ret[i][j] << "\t";
        }
        cout <<endl;
    }
    return 0;
}
